plugins/livelog/livelog.go

```diff
@@ -1,4 +1,125 @@
-//go:generate go-composite-schema --unexported --required config config-schema.yml generated_configschema.go
-
-// Package livelog is an awesome package
+// Package livelog implements a webhook handler for serving up livelogs of a task
+// sandbox.
 package livelog
+
+import (
+	"bufio"
+	"fmt"
+	"net/http"
+
+	"github.com/Sirupsen/logrus"
+	"github.com/taskcluster/taskcluster-client-go/tcclient"
+	"github.com/taskcluster/taskcluster-worker/plugins"
+	"github.com/taskcluster/taskcluster-worker/plugins/extpoints"
+	"github.com/taskcluster/taskcluster-worker/runtime"
+)
+
+type pluginProvider struct {
+	extpoints.PluginProviderBase
+}
+
+type plugin struct {
+	plugins.PluginBase
+	log *logrus.Entry
+}
+
+type taskPlugin struct {
+	plugins.TaskPluginBase
+	context    *runtime.TaskContext
+	url        string
+	expiration tcclient.Time
+	log        *logrus.Entry
+}
+
+func (pluginProvider) NewPlugin(opts extpoints.PluginOptions) (plugins.Plugin, error) {
+	return plugin{log: opts.Log}, nil
+}
+
+func (p plugin) NewTaskPlugin(opts plugins.TaskPluginOptions) (plugins.TaskPlugin, error) {
+	return &taskPlugin{
+		TaskPluginBase: plugins.TaskPluginBase{},
+		log: p.log.WithFields(logrus.Fields{
+			"taskID": opts.TaskInfo.TaskID,
+			"runID":  opts.TaskInfo.RunID,
+		}),
+	}, nil
+}
+
+func (tp *taskPlugin) Prepare(context *runtime.TaskContext) error {
+	tp.context = context
+
+	tp.url = tp.context.AttachWebHook(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		// TODO (garndt): add support for range headers.  Might not be used at all currently
+		logReader, err := tp.context.NewLogReader()
+		if err != nil {
+			w.WriteHeader(500)
+			w.Write([]byte("Error opening up live log"))
+		}
+
+		defer logReader.Close()
+
+		flusher := w.(http.Flusher)
+		reader := bufio.NewReader(logReader)
+
+		for {
+			line, isPrefix, err := reader.ReadLine()
+			if err != nil {
+				return
+			}
+			w.Write(line)
+			if !isPrefix {
+				w.Write([]byte("\r\n"))
+			}
+
+			flusher.Flush()
```

Comment:
Flushing every line seems expensive... we would have to turn of nagle... and if there is a lot logs this would be bad...

I propose that we flush every 500ms or something like that..